volatility/framework/symbols/windows/pdb.py

```diff
@@ -0,0 +1,291 @@
+                    potential_output_filename))
+                continue
+            finally:
+                # If something else failed, removed the symbol file so we don't pick it up in the future
+                if not data_written and os.path.exists(potential_output_filename):
+                    os.remove(potential_output_filename)
+                # Clear out all the temporary file if we constructed one
+                for filename in tmp_files:
+                    try:
+                        os.remove(filename)
+                    except PermissionError:
+                        vollog.warning("Temporary file could not be removed: {}".format(filename))
+        else:
+            vollog.warning("Cannot write downloaded symbols, please add the appropriate symbols"
+                           " or add/modify a symbols directory that is writable")
+
+    @classmethod
+    def pdbname_scan(cls,
+                     ctx: interfaces.context.ContextInterface,
+                     layer_name: str,
+                     page_size: int,
+                     pdb_names: List[bytes],
+                     progress_callback: constants.ProgressCallback = None,
+                     start: Optional[int] = None,
+                     end: Optional[int] = None) -> Generator[Dict[str, Optional[Union[bytes, str, int]]], None, None]:
+        """Scans through `layer_name` at `ctx` looking for RSDS headers that
+        indicate one of four common pdb kernel names (as listed in
+        `self.pdb_names`) and returns the tuple (GUID, age, pdb_name,
+        signature_offset, mz_offset)
+
+        .. note:: This is automagical and therefore not guaranteed to provide correct results.
+
+        The UI should always provide the user an opportunity to specify the
+        appropriate types and PDB values themselves
+        """
+        min_pfn = 0
+
+        if start is None:
+            start = ctx.layers[layer_name].minimum_address
+        if end is None:
+            end = ctx.layers[layer_name].maximum_address
+
+        for (GUID, age, pdb_name,
+             signature_offset) in ctx.layers[layer_name].scan(ctx,
+                                                              PdbSignatureScanner(pdb_names),
+                                                              progress_callback = progress_callback,
+                                                              sections = [(start, end - start)]):
+            mz_offset = None
+            sig_pfn = signature_offset // page_size
+
+            for i in range(sig_pfn, min_pfn, -1):
+                if not ctx.layers[layer_name].is_valid(i * page_size, 2):
+                    break
+
+                data = ctx.layers[layer_name].read(i * page_size, 2)
+                if data == b'MZ':
+                    mz_offset = i * page_size
+                    break
+            min_pfn = sig_pfn
```

Comment:
@ikelos it seems like the way min_pfn and sig_pfn are handled would prevent multiple signatures on the same page from being processed/checked. This would occur as line 243 would set min_pfn to sig_pfn and then if the next loop iteration generated the same sig_pfn (if another signature was on the same page), then the loop on 235 would never iterate as sig_pfn and min_pfn would be the same. Was the intention to prevent multiple signatures on the sam page? If so, what led to that decision?

Reply:
This is literally a copy/paste from what we had previously (which had been in the code base for four years, see commit https://github.com/volatilityfoundation/volatility3/commit/d6039fb899c3e2447a5902ebcaa52569523ef313), so I can look into these, but reviewing old code somewhat makes it difficult to tell whether the changes were acceptable or not.  5:S  Looks like the code originally came from @npetroni 's pdbscanner, so I'll let him answer that question, if that's ok?

Given a valid MZ header is unlikely to have multiple signatures in a page (since loaded MZ headers are always mapped to the start of a page), I'm not sure only returning a single signature is problematic?  If we're looking for MZs that have just been arbitrarily loaded into memory then this function may not be suitable, we can add something to the docstring if that will help?

Reply:
Re: Old code.. I wasn't too familar with this file before looking at this PR so I had to go beyond the code changed to really understand what affect(s) might be caused by it. That is where those comments came from.

Re: MZ headers.. that makes sense! Thanks!